osquery/filesystem/posix/fileops.cpp

```diff
@@ -0,0 +1,217 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <glob.h>
+#include <pwd.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <sys/stat.h>
+#include <sys/time.h>
+
+#include <boost/filesystem.hpp>
+#include <boost/optional.hpp>
+
+#include "osquery/core/process.h"
+#include "osquery/filesystem/fileops.h"
+
+namespace fs = boost::filesystem;
+
+namespace osquery {
+
+PlatformFile::PlatformFile(const std::string& path, int mode, int perms) {
+  int oflag = 0;
+  bool may_create = false;
+  bool check_existence = false;
+
+  if ((mode & PF_READ) == PF_READ && (mode & PF_WRITE) == PF_WRITE) {
+    oflag = O_RDWR;
+  } else if ((mode & PF_READ) == PF_READ) {
+    oflag = O_RDONLY;
+  } else if ((mode & PF_WRITE) == PF_WRITE) {
+    oflag = O_WRONLY;
+  }
+
+  switch (PF_GET_OPTIONS(mode)) {
+    case PF_GET_OPTIONS(PF_CREATE_ALWAYS):
+      oflag |= O_CREAT;
+      may_create = true;
+      break;
+    case PF_GET_OPTIONS(PF_CREATE_NEW):
+      oflag |= O_CREAT | O_EXCL;
+      may_create = true;
+      break;
+    case PF_GET_OPTIONS(PF_OPEN_EXISTING):
+      check_existence = true;
+      break;
+    case PF_GET_OPTIONS(PF_TRUNCATE):
+      if (mode & PF_WRITE) oflag |= O_TRUNC;
+      break;
+    default:
+      break;
+  }
+
+  if ((mode & PF_NONBLOCK) == PF_NONBLOCK) {
+    oflag |= O_NONBLOCK;
+    is_nonblock_ = true;
+  }
+
+  if (perms == -1 && may_create) {
+    perms = 0666;
+  }
+
+  if (check_existence && !fs::exists(path.c_str())) {
+    handle_ = kInvalidHandle;
+  } else {
+    handle_ = ::open(path.c_str(), oflag, perms);
+  }
+}
+
+PlatformFile::~PlatformFile() { 
+  if (handle_ != kInvalidHandle) {
+    ::close(handle_);
+  }
+}
+
+bool PlatformFile::isFile() const {
+  return (size() > 0);
+}
+
+bool PlatformFile::isOwnerRoot() const {
+  struct stat file;
+  if (fstat(handle_, &file) < 0) {
+    return false;
+  }
+
+  return (file.st_uid == 0);
+}
+
+bool PlatformFile::getFileTimes(PlatformTime& times) {
+  if (!isValid()) {
+    return false;
+  }
+
+  struct stat file;
+  if (::fstat(handle_, &file) < 0) {
```

Comment:
Inconsistent use of default namespace with the above `fstat`. I'd nix the `::` here.